### PR TITLE
[Data] Avoid `TypeError` when destructing `Dataset` object

### LINKED ISSUE #42382: [data] TypeError: ‘NoneType’ object is not callable error from ray data `map_batches`
### What happened + What you expected to happen

Observing a `TypeError` after completion of `map_batches` in ray data.
```py
Exception ignored in: <function Dataset.__del__ at 0x2aad7d3a0>
Traceback (most recent call last):
  File "/Users/arun/Applications/miniconda3/envs/dc/lib/python3.9/site-packages/ray/data/dataset.py", line 5148, in __del__
TypeError: 'NoneType' object is not callable
```

The job runs successfully, but it throws the above error which I am not able to catch as well. I think the error is specific to macOS because I was not able to reproduce the error in a linux machine.

### Versions / Dependencies

Python version: 3.9.17
Ray version: 2.9.0
OS: macOS 13.5 (M1 Chip)

### Reproduction script

Steps to reproduce:
1. Install deepchem (`pip install deepchem`)
2. Run the following code snippet:
```py
import deepchem as dc
import os
import ray

from typing import Dict, Any, Iterable, Optional, List
from functools import partial

# prepare data
import csv

with open('smiles.csv', 'w', newline='') as csvfile:
    fieldnames = ['smiles', 'logp']
    writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

    writer.writeheader()
    for i in range(1000):
        writer.writerow({'smiles': 'Cc1[nH]ccc1C(=O)N[C@@H](C)CNC(=O)c1cccc2[nH]ccc21', 'logp': 2})

ray.init(num_cpus=4)

def featurize(row: Dict[str, Any],
              featurizer,
              x='smiles',
              y='logp') -> Dict[str, Any]:
    return row


featurize_batches = partial(featurize, featurizer=dc.feat.CircularFingerprint())

# Featurizing a dataset
ds = ray.data.read_csv('smiles.csv').map_batches(featurize_batches)

# Writing a dataset to disk
ds.write_parquet('data-dir')

for i, batch in enumerate(ds.iter_batches()):
    print (i)

ray.shutdown()
```

The above code throws the following traceback:
```py
No normalization for AvgIpc. Feature removed!
Skipped loading some Jax models, missing a dependency. No module named 'jax'
2024-01-13 10:37:27,611	INFO worker.py:1715 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
2024-01-13 10:37:28,878	INFO set_read_parallelism.py:115 -- Using autodetected parallelism=8 for stage ReadCSV to satisfy parallelism at least twice the available number of CPUs (4).
2024-01-13 10:37:28,878	INFO set_read_parallelism.py:122 -- To satisfy the requested parallelism of 8, each read task output is split into 8 smaller blocks.
2024-01-13 10:37:28,878	INFO streaming_executor.py:112 -- Executing DAG InputDataBuffer[Input] -> TaskPoolMapOperator[ReadCSV] -> TaskPoolMapOperator[MapBatches(partial)->Write]
2024-01-13 10:37:28,879	INFO streaming_executor.py:113 -- Execution config: ExecutionOptions(resource_limits=ExecutionResources(cpu=None, gpu=None, object_store_memory=None), exclude_resources=ExecutionResources(cpu=0, gpu=0, object_store_memory=0), locality_with_output=False, preserve_order=False, actor_locality_enabled=True, verbose_progress=False)
2024-01-13 10:37:28,880	INFO streaming_executor.py:115 -- Tip: For detailed progress reporting, run `ray.data.DataContext.get_current().execution_options.verbose_progress = True`
(MapBatches(partial)->Write pid=1989) No normalization for AvgIpc. Feature removed!
(MapBatches(partial)->Write pid=1989) WARNING:tensorflow:From /Users/arun/Applications/miniconda3/envs/dc/lib/python3.9/site-packages/tensorflow/python/util/deprecation.py:576: calling function (from tensorflow.python.eager.polymorphic_function.polymorphic_function) with experimental_relax_shapes is deprecated and will be removed in a future version.
(MapBatches(partial)->Write pid=1989) Instructions for updating:                
(MapBatches(partial)->Write pid=1989) experimental_relax_shapes is deprecated, use reduce_retracing instead
2024-01-13 10:37:36,650 WARNING plan.py:588 -- Warning: The Ray cluster currently does not have any available CPUs. The Dataset job will hang unless more CPUs are freed up. A common reason is that cluster resources are used by Actors or Tune trials; see the following link for more details: https://docs.ray.io/en/latest/data/data-internals.html#ray-data-and-tune
2024-01-13 10:37:36,656	INFO set_read_parallelism.py:115 -- Using autodetected parallelism=8 for stage ReadCSV to satisfy parallelism at least twice the available number of CPUs (4).
2024-01-13 10:37:36,656	INFO set_read_parallelism.py:122 -- To satisfy the requested parallelism of 8, each read task output is split into 8 smaller blocks.
2024-01-13 10:37:36,656	INFO streaming_executor.py:112 -- Executing DAG InputDataBuffer[Input] -> TaskPoolMapOperator[ReadCSV] -> TaskPoolMapOperator[MapBatches(partial)]
2024-01-13 10:37:36,657	INFO streaming_executor.py:113 -- Execution config: ExecutionOptions(resource_limits=ExecutionResources(cpu=None, gpu=None, object_store_memory=None), exclude_resources=ExecutionResources(cpu=0, gpu=0, object_store_memory=0), locality_with_output=False, preserve_order=False, actor_locality_enabled=True, verbose_progress=False)
2024-01-13 10:37:36,657	INFO streaming_executor.py:115 -- Tip: For detailed progress reporting, run `ray.data.DataContext.get_current().execution_options.verbose_progress = True`
(MapBatches(partial)->Write pid=1989) Skipped loading some Jax models, missing a dependency. No module named 'jax'
(MapBatches(partial)->Write pid=1990) No normalization for AvgIpc. Feature removed! [repeated 3x across cluster] (Ray deduplicates logs by default. Set RAY_DEDUP_LOGS=0 to disable log deduplication, or see https://docs.ray.io/en/master/ray-observability/ray-logging.html#log-deduplication for more options.)
Running: 0.0/4.0 CPU, 0.0/0.0 GPU, 0.06 MiB/386.43 MiB object_store_memory:   0%/Users/arun/Applications/miniconda3/envs/dc/lib/python3.9/site-packages/ray/data/_internal/arrow_block.py:148: FutureWarning: promote has been superseded by mode='default'.
  return transform_pyarrow.concat(tables)
0                                                                               
1                                                                               
2
3
(MapBatches(partial)->Write pid=1990) WARNING:tensorflow:From /Users/arun/Applications/miniconda3/envs/dc/lib/python3.9/site-packages/tensorflow/python/util/deprecation.py:576: calling function (from tensorflow.python.eager.polymorphic_function.polymorphic_function) with experimental_relax_shapes is deprecated and will be removed in a future version. [repeated 3x across cluster]
(MapBatches(partial)->Write pid=1990) Instructions for updating: [repeated 3x across cluster]
(MapBatches(partial)->Write pid=1990) experimental_relax_shapes is deprecated, use reduce_retracing instead [repeated 3x across cluster]
(MapBatches(partial)->Write pid=1990) Skipped loading some Jax models, missing a dependency. No module named 'jax' [repeated 3x across cluster]
Exception ignored in: <function Dataset.__del__ at 0x2ac5019d0>
Traceback (most recent call last):
  File "/Users/arun/Applications/miniconda3/envs/dc/lib/python3.9/site-packages/ray/data/dataset.py", line 5148, in __del__
TypeError: 'NoneType' object is not callable
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

You might see a `TypeError` when your Ray Data program exits:

```
Exception ignored in: <function Dataset.__del__ at 0x2aad7d3a0>
Traceback (most recent call last):
  File "/Users/arun/Applications/miniconda3/envs/dc/lib/python3.9/site-packages/ray/data/dataset.py", line 5148, in __del__
TypeError: 'NoneType' object is not callable
```

The reason for this is that we don't correctly check if the `ray` module exists. Currently, we check if `ray is not None`:

https://github.com/ray-project/ray/blob/ba3cfa0eb0daaf9ecdb17edc1c4fa4bbeb4d9dd8/python/ray/data/dataset.py#L5062

But `ray` might evaluate to `<module None from None>`, which is not `None`. So, we evaluate `ray.is_initialized()` and get a `TypeError`. This PR fixes the issue by adding a more robust check.

## Related issue number

Fixes https://github.com/ray-project/ray/issues/42382
## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/dataset.py [hunk 1]
```diff
@@ -5059,7 +5059,18 @@ def __setstate__(self, state):
...
     def __del__(self):
-        if self._current_executor and ray is not None and ray.is_initialized():
+        if not self._current_executor:
+            return
+
+        # When Python shuts down, `ray` might evaluate to `<module None from None>`.
+        # This value is truthy and not `None`, so we use a try-catch rather than
+        # something like `if ray is not None`. For more information, see #42382.
+        try:
+            is_ray_initialized = ray.is_initialized()
```
Comment (omatthew98) [inline python/ray/data/dataset.py:5069]: Nit: Do we need to store the value i.e. would just the following work:
```
try:
    if ray.is_initialized():
        self._current_executor.shutdown()
except TypeError:
    pass
```

**Review by omatthew98 — APPROVED** [review]
One micro nit, but lgtm!

File: python/ray/data/dataset.py (hunk 1 above)
Comment (bveeramani) [reply python/ray/data/dataset.py]: Nah, that'd totally work. 

Updated..

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -5059,8 +5059,17 @@ def __setstate__(self, state):
...
-        if self._current_executor and ray is not None and ray.is_initialized():
-            self._current_executor.shutdown()
+        if not self._current_executor:
+            return
+
+        # When Python shuts down, `ray` might evaluate to `<module None from None>`.
+        # This value is truthy and not `None`, so we use a try-catch rather than
+        # something like `if ray is not None`. For more information, see #42382.
+        try:
+            if ray.is_initialized():
```
Comment (c21) [inline python/ray/data/dataset.py:5069]: we don't need to check `ray is not None` now?

**Review by c21 — APPROVED** [review]

File: python/ray/data/dataset.py (hunk 2 above)
Comment (bveeramani) [reply python/ray/data/dataset.py]: I don't think `ray` would ever be `None` (I suspect it'd always be `<module None from None>`), but I'll add it just to be safe.

Updated.

**Pull request merged**